pkg/updater/repository/repository.go

```diff
@@ -241,29 +315,111 @@ func movePackageFromSource(packageName string, rootPath string, sourcePath strin
+			isExperiment := r.experiment.Exists() && r.experiment.Target() == version.Name()
+			if isLink || isStable || isExperiment {
+				continue
+			}
+			log.Debugf("removing version %s", version.Name())
+			err := os.RemoveAll(filepath.Join(r.rootPath, version.Name()))
+			if err != nil {
+				return fmt.Errorf("could not remove version: %w", err)
+			}
+		}
+		return nil
+	}
+
+	// Else, we only delete the version if no processes are using it
+	versions, err := os.ReadDir(r.runPath)
 	if err != nil {
-		return fmt.Errorf("could not read root directory: %w", err)
+		return fmt.Errorf("could not read run directory: %w", err)
 	}
-	for _, file := range files {
-		isLink := file.Name() == stableVersionLink || file.Name() == experimentVersionLink
-		isStable := r.stable.Exists() && r.stable.Target() == file.Name()
-		isExperiment := r.experiment.Exists() && r.experiment.Target() == file.Name()
+
+	// For each version, get the PIDs
+	for _, version := range versions {
+		isLink := version.Name() == stableVersionLink || version.Name() == experimentVersionLink
+		isStable := r.stable.Exists() && r.stable.Target() == version.Name()
+		isExperiment := r.experiment.Exists() && r.experiment.Target() == version.Name()
 		if isLink || isStable || isExperiment {
 			continue
 		}
-		err := os.RemoveAll(filepath.Join(r.rootPath, file.Name()))
+
+		if !version.IsDir() {
+			continue
+		}
+
+		versionInUse, err := packageVersionInUse(filepath.Join(r.runPath, version.Name()))
 		if err != nil {
-			return fmt.Errorf("could not remove file: %w", err)
+			log.Errorf("could not check if package version is in use: %v", err)
+			continue
+		}
+
+		// If no PIDs are running, remove the version
+		if !versionInUse {
+			log.Debugf("no running PIDs for package %s version %s, removing package", r.rootPath, version.Name())
+			if err := os.RemoveAll(filepath.Join(r.rootPath, version.Name())); err != nil {
+				log.Errorf("could not remove package directory for version %s: %v", version, err)
+			}
+			if err := os.RemoveAll(filepath.Join(r.runPath, version.Name())); err != nil {
+				log.Errorf("could not remove run directory for version %s: %v", version, err)
+			}
 		}
 	}
+
 	return nil
 }
 
+// packageVersionInUse checks if the given package version is in use
+func packageVersionInUse(runPath string) (bool, error) {
+	runningPIDs := 0
+
+	pids, err := os.ReadDir(runPath)
+	if err != nil {
+		return false, fmt.Errorf("could not read run directory: %w", err)
+	}
+
+	// For each PID, check if it's running
+	for _, rawPID := range pids {
+		pid, err := strconv.ParseInt(rawPID.Name(), 10, 64)
+		if err != nil {
+			log.Errorf("could not parse PID: %v", err)
+			continue
+		}
+		process, err := os.FindProcess(int(pid))
+		if err != nil {
+			log.Errorf("could not find process with PID %d: %v", pid, err)
+		} else {
+			// Send a signal 0 to check if the process is running
+			if err := process.Signal(syscall.Signal(0)); err == nil {
```

Comment:
We should use `process.FindProcess`. Signals can always have fun side effects (it's a common pattern to trap on 0 to do operation on-exit)